### DISCUSSION TITLE
Using drift as a SQL query constructor

### DISCUSSION THREAD

**Discussion by BananaMasterz** [simolus3/drift#2512] [General]
I was wondering if one could use drift only to construct queries for a `MariaDB` database for example which is not `sqlite`, because drift is type safe and way better than writing raw queries that return untyped data.

I know drift is not meant to be used like that obviously, but would there be any obvious issues?

**Comment by simolus3** [comment]
Yes, you can implement a custom `QueryExecutor` (or, perhaps simpler, a `DatabaseDelegate` from `package:drift/backends.dart`) that makes drift talk to a MariaDB database.

> I know drift is not meant to be used like that obviously, but would there be any obvious issues?

Most parts of drift are built with the assumption of running against sqlite3. For DQL statements, the differences usually aren't that bad (although things like date and time functions probably won't work), but types or things like `WITHOUT ROWID` in table definitions will likely not work without further work on your side.

It think it can make sense to try this out, but be aware that you won't get a reliable solution like the one for sqlite3. If you're willing to tackle issues as they appear (PRs to add support for different DBMS implementations are welcome!), you can use drift with MariaDB.

**Reply by BananaMasterz** [reply]
I love drift and how easy it makes things so ofc I'd be very interested in trying to make it work with MariaDB if you say it's possible.

I thought it wasn't gonna be easy cuz you can't even connect with username and password in SQLite. The connection string for MariaDB should be supported etc... That's why I only asked regarding the query construction.

Would you say then it's worth trying to make drift work/talk straight with MariaDB?

**Reply by simolus3** [reply]
The "lowest" layer of drift responsible for talking to the database (that is, opening a connection, sending SQL statements with arguments and interpreting results) can be replaced with a custom implementation. The interface describing the protocol is part of a public API, it's not internal to drift.

For instance, there's a package to make drift talk to a postgres database [here](https://github.com/simolus3/drift/blob/develop/extras/drift_postgres/lib/src/pg_database.dart), which includes support for authentication (as handled by the postgres package). Something similar should work for MariaDB and is likely easier to set up than generating SQL only and then handling the rest outside of a drift API.

**Reply by BananaMasterz** [reply]
Using [mysql_client](https://pub.dev/packages?q=mysql_client) package I managed to get drift to work with MariaDB based on the postgres example you linked, but I did it with some very junky workarounds (just to make it work as a first step).

One of the issues I encountered is `GenerationContext.sql` returns a string with double quotes which is not valid in mariadb, it insteads uses backtick ``` ` ```. The junky workaround I did was go inside the `generation_context.dart` file and do
```String get sql => buffer.toString().replaceAll("\"", "`");```
similarly the parameters in where coming out as `$1` instead of `(?)` so I did 
```String get sql => buffer.toString().replaceAll("\"", "`").replaceAll("\$1", "(?)");```
but I'm quite sure there's a way better way to fix both of these. I've never been this deep into drift so I have absolutely no idea what I'm doing lol

Finally one thing that I can't figure out at all is how to implement a function I used to have that looked like this:
```dart
NativeDatabase.createInBackground(file, setup: (rawDb) {
  rawDb.createFunction(
    functionName: 'removeDiacritics',
    function: (args) {
      final text = args[0] as String?;
      return text != null ? removeDiacritics(text) : null;
    },
  );
});
```
but now I'm not using `NativeDatabase` obviously so not sure how to go about doing this. I think it should be simple I just haven't figured it out.

Here's my currently edited code based on the above postgress implementation (obviously haven't worked on the types yet since I'm far from understanding those yet so I just went with the example's PgTypedParameter that somehow seem to be okish)

<details><summary>Show code</summary>
<p>

```dart
import 'dart:async';

import 'package:collection/collection.dart';
import 'package:drift/backends.dart';
import 'package:mysql_client/mysql_client.dart';
import 'package:postgres/postgres_v3_experimental.dart';

class MariaDatabase extends DelegatedDatabase {
  MariaDatabase({
    bool logStatements = false,
  }) : super(
          _MariaDelegate(
            () => MySQLConnectionPool(
              host: 'localhost',
              port: 3306,
              userName: 'root',
              password: r"....",
              databaseName: 'my_database_name',
              maxConnections: 10,
              secure: false,
            ),
            true,
          ),
          isSequential: true,
          logStatements: logStatements,
        );

  /// Creates a drift database implementation from a postgres database
  /// [connection].
  MariaDatabase.opened(MySQLConnectionPool connection, {bool logStatements = false})
      : super(
          _MariaDelegate(() => connection, false),
          isSequential: true,
          logStatements: logStatements,
        );

  @override
  SqlDialect get dialect => SqlDialect.sqlite;
}

class _MariaDelegate extends DatabaseDelegate {
  _MariaDelegate(this._open, this.closeUnderlyingWhenClosed);

  final bool closeUnderlyingWhenClosed;
  final FutureOr<MySQLConnectionPool> Function() _open;

  MySQLConnectionPool? _openedSession;

  @override
  TransactionDelegate get transactionDelegate => const NoTransactionDelegate();

  @override
  late DbVersionDelegate versionDelegate;

  @override
  Future<bool> get isOpen => Future.value(_openedSession != null);

  @override
  Future<void> open(QueryExecutorUser user) async {
    final session = await _open();
    final mariaVersionDelegate = _MariaVersionDelegate(session);

    await mariaVersionDelegate.init();

    _openedSession = session;
    versionDelegate = mariaVersionDelegate;
  }

  @override
  Future<void> runBatched(BatchedStatements statements) async {
    final session = _openedSession!;
    final prepared = List<PreparedStmt?>.filled(statements.statements.length, null);

    try {
      for (final instantation in statements.arguments) {
        final mariaArgs = _BoundArguments.ofDartArgs(instantation.arguments);

        // Lazily prepare statements when we run into them. The reason is that
        // we need to know the types for variables.
        final stmtIndex = instantation.statementIndex;
        var stmt = prepared[stmtIndex];
        if (stmt == null) {
          final sql = statements.statements[stmtIndex];
          stmt = prepared[stmtIndex] = await session.prepare(sql);
        }

        await stmt.execute(mariaArgs.parameters);
      }
    } finally {
      for (final stmt in prepared) {
        await stmt?.deallocate();
      }
    }
  }

  Future<int> _runWithArgs(String statement, List<Object?> args) async {
    final session = _openedSession!;

    final mariaArgs = _BoundArguments.ofDartArgs(args);
    var stmt = await session.prepare(statement);
    final result = await stmt.execute(mariaArgs.parameters);
    return result.affectedRows.toInt();
  }

  @override
  Future<void> runCustom(String statement, List<Object?> args) async {
    await _runWithArgs(statement, args);
  }

  @override
  Future<int> runInsert(String statement, List<Object?> args) async {
    final session = _openedSession!;
    final mariaArgs = _BoundArguments.ofDartArgs(args);
    var stmt = await session.prepare(statement);
    final result = await stmt.execute(mariaArgs.parameters);
    return result.firstOrNull?.first as int? ?? 0;
  }

  @override
  Future<int> runUpdate(String statement, List<Object?> args) async {
    return _runWithArgs(statement, args);
  }

  @override
  Future<QueryResult> runSelect(String statement, List<Object?> args) async {
    final session = _openedSession!;
    final mariaArgs = _BoundArguments.ofDartArgs(args);
    var stmt = await session.prepare(statement);
    var parsTransformed = List.generate(mariaArgs.parameters.length, (index) => mariaArgs.parameters[index].value);

    final result = await stmt.execute(parsTransformed);

    var rowsList = result.rows.toList();
    var rawRowValues = List.generate(rowsList.length, (index) => rowsList[index].typedAssoc().values.toList());

    return QueryResult([
      for (final mariaColumn in result.cols) mariaColumn.name ?? '',
    ], rawRowValues);
  }

  @override
  Future<void> close() async {
    if (closeUnderlyingWhenClosed) {
      await _openedSession?.close();
    }
  }
}

class _BoundArguments {
  final List<PgDataType> types;
  final List<PgTypedParameter> parameters;

  _BoundArguments(this.types, this.parameters);

  factory _BoundArguments.ofDartArgs(List<Object?> args) {
    final types = <PgDataType>[];
    final parameters = <PgTypedParameter>[];

    void add(PgTypedParameter param) {
      types.add(param.type);
      parameters.add(param);
    }

    for (final value in args) {
      if (value == null) {
        add(PgTypedParameter(PgDataType.text, null));
      } else if (value is int) {
        add(PgTypedParameter(PgDataType.bigInteger, value));
      } else if (value is BigInt) {
        // Drift only uses BigInts to represent 64-bit values on the web, so we
        // can use toInt() here.
        add(PgTypedParameter(PgDataType.bigInteger, value));
      } else if (value is bool) {
        add(PgTypedParameter(PgDataType.boolean, value));
      } else if (value is double) {
        add(PgTypedParameter(PgDataType.double, value));
      } else if (value is String) {
        add(PgTypedParameter(PgDataType.text, value));
      } else if (value is List<int>) {
        add(PgTypedParameter(PgDataType.byteArray, value));
      } else {
        throw ArgumentError.value(value, 'value', 'Unsupported type');
      }
    }

    return _BoundArguments(types, parameters);
  }
}

class _MariaVersionDelegate extends DynamicVersionDelegate {
  final MySQLConnectionPool database;

  _MariaVersionDelegate(this.database);

  @override
  Future<int> get schemaVersion async {
    final result = await database.execute('SELECT version FROM __schema');
    return int.parse(result.rows.first.colAt(0)!) as int;
  }

  Future init() async {
    await database.execute('CREATE TABLE IF NOT EXISTS __schema (version integer NOT NULL DEFAULT 0)');

    final count = await database.execute('SELECT COUNT(*) FROM __schema');
    if (int.parse(count.rows.first.colAt(0)!) as int == 0) {
      await database.execute('INSERT INTO __schema (version) VALUES (0)');
    }
  }

  @override
  Future<void> setSchemaVersion(int version) async {
    var stmt = await database.prepare(r'UPDATE __schema SET version = (?)');
    await stmt.execute([version]);
  }
}
```

</p>
</details>

**Reply by BananaMasterz** [reply]
**discard my previous reply ( i deleted it) cuz realized I was accidentally using `assoc()` instead of `typedAssoc()` on the results of the query execution which fixed a lot of things

**Reply by BananaMasterz** [reply]
so I've been trying to figure out transactions and I set 
```dart
  @override
  TransactionDelegate get transactionDelegate => const NoTransactionDelegate(
        start: 'START TRANSACTION',
      );
```
in my `DatabaseDelegate` since mariadb uses `START` instead of `BEGIN` but for some reason it hangs in `DatabaseConnectionUser`'s `_runConnectionZoned` and I have no idea why. It doesn't throw an error or anything, it just never returns. More specifically it hangs here:

```dart
Future<T> transaction<T>(Future<T> Function() action,
    {bool requireNew = false}) async {
  final resolved = resolvedEngine;

  // Are we about to start a nested transaction?
  if (resolved is Transaction) {
    final executor = resolved.executor as TransactionExecutor;
    if (!executor.supportsNestedTransactions) {
      if (requireNew) {
        throw UnsupportedError('The current database implementation does '
            'not support nested transactions.');
      } else {
        // Just run the block in the current transaction zone.
        return action();
      }
    }
  }

  return await resolved.doWhenOpened((executor) {
    final transactionExecutor = executor.beginTransaction();
    final transaction = Transaction(this, transactionExecutor);

    return _runConnectionZoned(transaction, () async {
      print('print tr 1');
      var success = false;
      try {
        print('print tr 2');
        final result = await action();
        print('print tr 3');
        success = true;
        return result;
      } catch (e, s) {
        print('print tr 4');
        await transactionExecutor.rollbackAfterException(e, s);
        // pass the exception on to the one who called transaction()
        rethrow;
      } finally {
        print('print tr 5');
        if (success) {
          try {
            await transaction.complete();
          } catch (e, s) {
            // Couldn't commit -> roll back then.
            await transactionExecutor.rollbackAfterException(e, s);
            rethrow;
          }
        }
        await transaction.disposeChildStreams();
      }
    });
  });
}
```
`print tr 2` is printed but neither 4 nor 5 are, which means it hangs in `final result = await action();` what could a possible reason for this be?

**Reply by BananaMasterz** [reply]
idk why I wasn't getting any errors, but I solved the issue. Turns out `_runWithArgs` was executing `runCustom` with parameters and `START TRANSACTION` doesn't have args so the sql package I'm using doesn't handle it as I expected. Instead I had to do this to handle the case of no empty args:
```dart
Future<int> _runWithArgs(String statement, List<Object?> args) async {
  final session = _openedSession!;
  var mariaArgs = _BoundArguments.ofDartArgs(args);
  IResultSet result;
  if (args.isEmpty) {
    result = await session.execute(statement);
  } else {
    var stmt = await session.prepare(statement);
    result = await stmt.execute(mariaArgs.parameters);
  }
  return result.affectedRows.toInt();
}
```
also updated my `TransactionDelegate` accordingly 
```dart
@override
TransactionDelegate get transactionDelegate => const NoTransactionDelegate(
      start: 'START TRANSACTION',
      commit: 'COMMIT',
      rollback: 'ROLLBACK',
    );
```

finally, for nested transactions, I had to edit the `_StatementBasedTransactionExecutor.nested` the following way:
```dart
_StatementBasedTransactionExecutor.nested(
    _StatementBasedTransactionExecutor this._parent, int depth)
    : _delegate = _parent._delegate,
      _startCommand = 'SAVEPOINT s$depth',
      _commitCommand = 'RELEASE SAVEPOINT s$depth',
      _rollbackCommand = 'ROLLBACK TO SAVEPOINT s$depth',
      super(_parent._db);
```
basically changed
from `'RELEASE s$depth'` to `'RELEASE SAVEPOINT s$depth'`
and 
from `'ROLLBACK TO s$depth'` to `'ROLLBACK TO SAVEPOINT s$depth'`
to make up for slight deviation of SqlLite and MariaDB syntax.
Maybe this could be done on a higher level like the `TransactionDelegate` but it probably needs more work so for now I'll leave it as is. Might change it based on `SqlDialect` by introducing `MariaDB` dialect and using that to set these vars.

Now I have a fully working app that talks to MariaDB using drift. Absolutely amazing!! Only thing missing (I think) are the functions that I haven't figured out yet as per my earlier reply.

**Reply by simolus3** [reply]
Sorry for the delayed response, this looks like awesome work on your part!! All of the changes you've made (even if you think they're incomplete) would be very welcome in a PR to drift :) They're a great starting point, and I'm happy to help out with this work too.

> Finally one thing that I can't figure out at all is how to implement a function I used to have that looked like this:

Unfortunately, creating a function in MariaDB is [a lot more complicated](https://mariadb.com/kb/en/user-defined-functions/) since you need to write it in C and compile it into a shared library loaded by the server process. Maybe the better collate support in MariaDB can be used as an alternative?

**Reply by BananaMasterz** [reply]
No worries. You've done some great work with drift and that's the only reason I got this far with MariaDB! Will absolutely share more once I test a bit more stuff.

Currently building features alongside these changes to see if anything comes up. Gotta do datetimes too (although the current implementation works as an int, idk if it's "right" to store in a MariaDB datetimes as int lol) also gotta test the migration next.

**Reply by BananaMasterz** [reply]
I introduced a new value in `SqlDialect`  named `mariadb` and in `column_constraints.dart` at `defaultConstraints` i want to replace `AUTOINCREMENT` with `AUTO_INCREMENT` depending on the dialect at this part of the code:
```dart
for (final feature in column.constraints) {
  if (feature is PrimaryKeyColumn) {
    if (!wrotePkConstraint) {
      defaultConstraints.add(feature.isAutoIncrement
          ? 'PRIMARY KEY AUTOINCREMENT'
          : 'PRIMARY KEY');

      wrotePkConstraint = true;
      break;
    }
  }
}
```

how could I gain access to `dialect` inside `defaultConstraints` to do something like this?
```dart
var autoIncrementWord = dialect == SqlDialect.sqlite ? 'AUTOINCREMENT' : dialect == SqlDialect.mariadb
    ? 'AUTO_INCREMENT' : 'AUTO_INCREMENT';

for (final feature in column.constraints) {
  if (feature is PrimaryKeyColumn) {
    if (!wrotePkConstraint) {
      defaultConstraints.add(feature.isAutoIncrement
          ? 'PRIMARY KEY $autoIncrementWord'
          : 'PRIMARY KEY');

      wrotePkConstraint = true;
      break;
    }
  }
}
```
unless ofc there's a better higher level location to do this.

I've also added `'AUTO_INCREMENT': TokenType.autoincrement,` in `Map<String, TokenType> keywords` of `token.dart`, is that correct?

**Reply by simolus3** [reply]
> how could I gain access to dialect inside defaultConstraints to do something like this?

You could try adding it to `dialectSpecificConstraints`, possibly like this:

```dart
      if (!wrotePkConstraint) {
        for (final dialect in SqlDialect.values) {
          if (dialect == SqlDialect.mysql) {
            dialectSpecificConstraints[dialect]!
                .add('PRIMARY KEY AUTO_INCREMENT');
          } else {
            dialectSpecificConstraints[dialect]!
                .add('PRIMARY KEY AUTOINCREMENT');
          }
        }
```

The generated code will include the differences for every dialect, so we can select the right one depending on the database at runtime.

> I've also added 'AUTO_INCREMENT': TokenType.autoincrement, in Map<String, TokenType> keywords of token.dart, is that correct?

This is only used for parsing SQL during the build step, e.g. for [drift files](https://drift.simonbinder.eu/docs/using-sql/drift_files/). I don't think it's necessary to support MariaDB syntax for that, at the moment the parser is pretty much sqlite-specific. It would be great if we could eventually translate table definitions in drift files (so that a `AUTOINCREMENT` column in a drift file would automatically become `AUTO_INCREMENT`), but we probably won't need that initially.

**Reply by BananaMasterz** [reply]
> You could try adding it to dialectSpecificConstraints, possibly like this:

Would I do this on top of the exiting `defaultConstraints`? For example would it be like this:
```dart

  for (final feature in column.constraints) {
    if (feature is PrimaryKeyColumn) {
      if (!wrotePkConstraint) {
        defaultConstraints.add(feature.isAutoIncrement
            ? 'PRIMARY KEY AUTOINCREMENT'
            : 'PRIMARY KEY');

        wrotePkConstraint = true;
        break;
      }
    }
  }

  if (!wrotePkConstraint) {
    for (final dialect in SqlDialect.values) {
      if (dialect == SqlDialect.mysql) {
        dialectSpecificConstraints[dialect]!
            .add('PRIMARY KEY AUTO_INCREMENT');
      } else {
        dialectSpecificConstraints[dialect]!
            .add('PRIMARY KEY AUTOINCREMENT');
      }
    }
```
Or would the first `for` be completely replaced by this? Sorry if it's kind of a dumb question but I'm not sure since it seems to be deliberately added in `defaultConstraints` and I'm not sure what that could entail...

> This is only used for parsing SQL during the build step, e.g. for [drift files](https://drift.simonbinder.eu/docs/using-sql/drift_files/).

Ah that's great then!

Also do you think there would be an easy way to:
1) have different `baseKeywords` depending on dialect?
2) have different escape characters depending on dialect? For example in drift `Column` the ```String get escapedName => '`$name`';``` instead of double quotes used in SQLite (same for `extension NameWithAlias` and in `GenerationContext` at 
```String identifier(String columnName) => '`$columnName`';```)

**Reply by BananaMasterz** [reply]
another thing I did is about datetimes, I am again not sure how to do this depending on dialect
```dart
const Expression<DateTime> currentDate = _DependingOnDateTimeExpression(
  forIsoString: _currentDateLiteral,
  forTimestamps: FunctionCallExpression(
    'UNIX_TIMESTAMP',
    [],
  ),
);


const Expression<DateTime> currentDateAndTime = _DependingOnDateTimeExpression(
  forIsoString: _currentTimestampLiteral,
  forTimestamps: FunctionCallExpression(
    'UNIX_TIMESTAMP',
    [],
  ),
);
```
these are for mariadb for example, although not sure if that's exactly what these functions do (i assumed they just get the current timestamp in SQLite same way this does, since it was using `%s` for SQLite)

**Reply by simolus3** [reply]
> Sorry if it's kind of a dumb question but I'm not sure since it seems to be deliberately added in defaultConstraints and I'm not sure what that could entail...

Not a dumb question at all! At the moment we're adding it to `defaultConstraints` because all dialects we currently support use `AUTOINCREMENT`. Once we add support for MariaDb, it can no longer be in `defaultConstraints` because it's now dialect-specific.

> have different escape characters depending on dialect?

Hm yeah, at runtime we can probably do it by looking up the associated dialect in the `GenerationContext`. We're also doing it at compile time in some instances, I'll need to come up with something for that.
At runtime, we can probably just add the MariaDb keywords unconditionally - it doesn't really hurt if additional identifiers get escaped.

> another thing I did is about datetimes, I am again not sure how to do this depending on dialect

I think we currently don't support datetimes with postgres either - these functions are very sqlite3-specific. Do you know if `CURRENT_TIME` works in MariaDb? If it does, you could probably just use enable [dates as text](https://drift.simonbinder.eu/docs/getting-started/advanced_dart_tables/#datetime-options) for `currentDate` to work.

**Reply by BananaMasterz** [reply]
> Not a dumb question at all! At the moment we're adding it to defaultConstraints because all dialects we currently support use AUTOINCREMENT. Once we add support for MariaDb, it can no longer be in defaultConstraints because it's now dialect-specific.

Alright, just to make sure then, it will become like this:
```dart
  for (final feature in column.constraints) {
    if (feature is PrimaryKeyColumn) {
      if (!wrotePkConstraint) {
        wrotePkConstraint = true;
        break;
      }
    }
  }

  if (!wrotePkConstraint) {
    for (final dialect in SqlDialect.values) {
      if (dialect == SqlDialect.mariadb) {
        dialectSpecificConstraints[dialect]!
            .add('PRIMARY KEY AUTO_INCREMENT');
      } else {
        dialectSpecificConstraints[dialect]!
            .add('PRIMARY KEY AUTOINCREMENT');
      }
    }
  }
  ```
  should I completely remove the empty `for` now since it does nothing or should I leave it there in case it's needed in the future?
  
  > Hm yeah, at runtime we can probably do it by looking up the associated dialect in the GenerationContext. We're also doing it at compile time in some instances
  
Yeah so far **I think** the only place we don't have access to `dialect` is on the `escapedName` of  `Column`

> Do you know if CURRENT_TIME works in MariaDb? If it does, you could probably just use enable [dates as text](https://drift.simonbinder.eu/docs/getting-started/advanced_dart_tables/#datetime-options) for currentDate to work.

`CURRENT_TIME` works in MariaDB but I prefer having the datetime as interger instead of text, also would `dates as text` also work for datetime if I were to enable it?

**Reply by simolus3** [reply]
> should I completely remove the empty for now since it does nothing or should I leave it there in case it's needed in the future?

Sorry, no. The for loop iterating over dialects should be moved into the original for loop, replacing the existing code adding to `defaultConstraints`.

> Yeah so far I think the only place we don't have access to dialect is on the escapedName of Column

Hm, maybe that would have to be refactored to be a function then.

> CURRENT_TIME works in MariaDB but I prefer having the datetime as interger instead of text

`CURRENT_TIME` returns a textual representation though, right?

**Reply by BananaMasterz** [reply]
> Sorry, no. The for loop iterating over dialects should be moved into the original for loop, replacing the existing code adding to defaultConstraints.

Ahh ok, my bad. Should be like this then:
```dart
for (final feature in column.constraints) {
  if (feature is PrimaryKeyColumn) {
    if (!wrotePkConstraint) {
      for (final dialect in SqlDialect.values) {
        if (dialect == SqlDialect.mariadb) {
          dialectSpecificConstraints[dialect]!
              .add('PRIMARY KEY AUTO_INCREMENT');
        } else {
          dialectSpecificConstraints[dialect]!
              .add('PRIMARY KEY AUTOINCREMENT');
        }
      }
      wrotePkConstraint = true;
      break;
    }
  }
}
```
> Hm, maybe that would have to be refactored to be a function then.

I don't think I can do that 😅 For now I leave it hardcoded with backtick while I'm testing other things as well unless of course it's something that would be easy for you.

> CURRENT_TIME returns a textual representation though, right?

If I understand your question correctly, yes as seen [here](https://mariadb.com/kb/en/curtime/) (`CURRENT_TIME` and `CURRENT_TIME()` are synonyms for `CURTIME()`)

By the way I have shipped migration to MariaDB from SQLite successfully to my production app (with the changes I've mentioned above hardcoded plus some other minor stuff I will share in a pull request as well when more testing is done) and it works extremely well. I haven't had a single issue reported. Great work by you for building such good base!

**Reply by simolus3** [reply]
> For now I leave it hardcoded with backtick while I'm testing other things as well unless of course it's something that would be easy for you.

It's probably not easy and will require a bit of refactoring :D Feel free to push the hardcoded changes so I know where to look, thanks!

**Reply by BananaMasterz** [reply]
> Feel free to push the hardcoded changes so I know where to look

I have pushed the changes (I **hope** I didn't miss any, if I did it will hopefully come up sooner or later while I'm running lots of complicated queries, if it doesn't somebody else might find it 😅) as you can see, only in one instance we don't have access to the dialect, at 
`Column.escapedName`

**Reply by BananaMasterz** [reply]
In mariadb, the following query is not valid:
```DELETE FROM `project` WHERE `project_id` IN () AND `user_id` = 15;```
The reason is `IN ()` in an empty list throws syntax error. The error I get in my flutter app is:

```
You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ') AND `user_id` = ?' at line 1
```

and I get no information as to where the error comes from in order to handle it as I want (for example I won't run the query if the Iterable I pass in `isIn` is empty.

Do you think it would be a good idea to do something like this?
```dart
Expression<bool> isIn(Iterable<D> values) {
  if (values.isEmpty) {
    throw 'tried to use "isIn" with empty iterable';
  }
  return _InExpression(this, values.toList(), false);
}

Expression<bool> isNotIn(Iterable<D> values) {
  if (values.isEmpty) {
    throw 'tried to use "isNotIn" with empty iterable';
  }
  return _InExpression(this, values.toList(), true);
}
```
We do not have access to dialect here but maybe it could be okay for SQLite too? Tho from what I understand SQLite doesn't care if it's empty. This way you know where the error is coming from and you can check your Iterable before doing something. Do you have any other solution or other location to implement something similar?

**Reply by simolus3** [reply]
>  Tho from what I understand SQLite doesn't care if it's empty

Empty iterables work as expected in sqlite3 though, `3 IN ()` evaluates to `0` and `3 NOT IN ()` evaluates to `1`. Perhaps we can return `const Constant(false)` in `isIn` for empty iterables and `const Constant(true)` in `isNotIn`? That way we preserve the current behavior.

**Reply by BananaMasterz** [reply]
ah you are absolutely right... no idea why I didn't think of (well cuz I didn't think `Constant(false)` is an `Expression<bool>` type required by the return type oops... yeah that'll do it. This is great then:
```dart
Expression<bool> isIn(Iterable<D> values) {
  if (values.isEmpty) { // mariadb dialect
    return Constant(false);
  }
  return _InExpression(this, values.toList(), false);
}

Expression<bool> isNotIn(Iterable<D> values) {
  if (values.isEmpty) {
    return Constant(true);
  }
  return _InExpression(this, values.toList(), true);
}
```
Would this also be ok for Sqlite then?

**Reply by simolus3** [reply]
I think it's ok. I've added tests verifying the current behavior in e98da9cecb9c4f3fcf177e8e9a32971d9de88f93. If those tests still pass with your changes, we're good :smile:
